### PR TITLE
Fix repeatedly initialize shared memory data and protect the memory's fields

### PR DESCRIPTION
**Pull request by wenyongh**
Add shared memory lock when accessing some fields of the memory instance
when the memory instance is shared.
And avoid repeatedly initializing the shared memory data when creating the
child thread in lib-pthread or lib-wasi-threads.

### CODE REVIEW

**Review by xujuntwt95329 — APPROVED** [review]
LGTM

File: core/iwasm/interpreter/wasm_runtime.c [hunk 1]
```diff
@@ -1774,12 +1774,31 @@ wasm_instantiate(WASMModule *module, WASMModuleInstance *parent,
...
 #if WASM_ENABLE_BULK_MEMORY != 0
         if (data_seg->is_passive)
             continue;
 #endif
 
+#if WASM_ENABLE_SHARED_MEMORY != 0
+        /* Currently we have only one memory instance */
+        if (module->import_memory_count > 0) {
+            is_shared_memory =
+                module->import_memories[0].u.memory.flags ? true : false;
```
Comment (loganek) [inline core/iwasm/interpreter/wasm_runtime.c:1790]: Should we check for the 0x2 flag here as well? Or imported memory cannot have any other flag values other than 0x2?

Comment (wenyongh) [reply core/iwasm/interpreter/wasm_runtime.c]: Yes, it is a mistake, thanks a lot for spotting, fixed.

**Review by loganek — APPROVED** [review]
LGTM, just a question on the performance - I see that adds extra lock/unlock for all WASI calls. Is there any other significant scenario that will be affected by this change from perf perspective?

**Comment by wenyongh** [comment]
> LGTM, just a question on the performance - I see that adds extra lock/unlock for all WASI calls. Is there any other significant scenario that will be affected by this change from perf perspective?

Yes, performance may be an issue, we had better test more. Maybe we can refine the code, e.g., add lock/unlock only when the memory is actually shared, and load/store memory's memory_data_size/memory_data_end fields with atomic operations, like what we did for suspend_flags.

**Comment by wenyongh** [comment]
> LGTM, just a question on the performance - I see that adds extra lock/unlock for all WASI calls. Is there any other significant scenario that will be affected by this change from perf perspective?

https://github.com/bytecodealliance/wasm-micro-runtime/pull/2682

**Comment by wenyongh** [comment]
@loganek I have merged #2682, if there is any issue found or there is enhancement needed, let's create another PR to fix/enhance it. Thanks.

**Comment by loganek** [comment]
@wenyongh thanks a lot and sorry for a slow response, I just came back from holidays.

**Pull request merged**